src/box/xrow.c

```diff
@@ -1921,21 +1921,39 @@ xrow_encode_join(struct xrow_header *row, const struct tt_uuid *instance_uuid)
 }
 
 int
-xrow_encode_vclock(struct xrow_header *row, const struct vclock *vclock)
+xrow_encode_heartbeat(struct xrow_header *row, uint32_t replica_id, double tm,
+		      const struct vclock *vclock, uint64_t vclock_sync)
 {
 	memset(row, 0, sizeof(*row));
-
-	/* Add vclock to response body */
-	size_t size = 8 + mp_sizeof_vclock_ignore0(vclock);
+	row->type = IPROTO_OK;
+	row->replica_id = replica_id;
+	row->tm = tm;
+	/* Reserve mp_sizeof(map + vclock(key) + vclock_sync(key + value)). */
+	size_t size = 12;
```

Comment:
Let's compute this number instead of using a magic constant (12).

Reply:
Ok:
```diff
diff --git a/src/box/xrow.c b/src/box/xrow.c
index 95d7fe369..24398af2a 100644
--- a/src/box/xrow.c
+++ b/src/box/xrow.c
@@ -1926,16 +1926,21 @@ xrow_encode_heartbeat(struct xrow_header *row, const struct vclock *vclock,
 {
        memset(row, 0, sizeof(*row));
        row->type = IPROTO_OK;
-       /* Reserve mp_sizeof(map + vclock(key) + vclock_sync(key + value)). */
-       size_t size = 12;
+       size_t size = 0;
        size_t map_size = 0;
        if (vclock != NULL) {
                map_size += 1;
+               size += mp_sizeof_uint(IPROTO_VCLOCK);
                size += mp_sizeof_vclock_ignore0(vclock);
        }
-       map_size += vclock_sync != 0;
+       if (vclock_sync != 0) {
+               map_size += 1;
+               size += mp_sizeof_uint(IPROTO_VCLOCK_SYNC);
+               size += mp_sizeof_uint(vclock_sync);
+       }
        if (map_size == 0)
                return 0;
+       size += mp_sizeof_map(map_size);
        char *buf = (char *) region_alloc(&fiber()->gc, size);
        if (buf == NULL) {
                diag_set(OutOfMemory, size, "region_alloc",
```